File: riot-api/src/main/java/fr/nspu/riot_api/RiotError.java
package fr.nspu.riot_api;

/**
 * Created by nspu on 09/03/18.
 */

import fr.nspu.riot_api.models.ErrorDetails;
import fr.nspu.riot_api.models.ErrorResponse;
import retrofit.RetrofitError;

/**
 * This object wraps error responses from the API
 * and provides access to details returned by the request that are usually more
 * descriptive than default Retrofit error messages.
 * <p>
 * To use with asynchronous requests pass {@link RiotCallback}
 * instead of {@link retrofit.Callback} when making the request:
 */
public class RiotError extends Exception {

    private final RetrofitError mRetrofitError;
    private final ErrorDetails mErrorDetails;

    public static RiotError fromRetrofitError(RetrofitError error) {
        ErrorResponse errorResponse = null;

        try {
            errorResponse = (ErrorResponse) error.getBodyAs(ErrorResponse.class);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (errorResponse != null && errorResponse.error != null) {
            String message = errorResponse.error.status + " " + errorResponse.error.message;
            return new RiotError(error, errorResponse.error, message);
        } else {
            return new RiotError(error);
        }
    }

    public RiotError(RetrofitError retrofitError, ErrorDetails errorDetails, String message) {
        super(message, retrofitError);
        mRetrofitError = retrofitError;
        mErrorDetails = errorDetails;
    }

    public RiotError(RetrofitError retrofitError) {
        super(retrofitError);
        mRetrofitError = retrofitError;
        mErrorDetails = null;
    }

    /**
     * @return the original {@link retrofit.RetrofitError} that was returned for this request.
     */
    public RetrofitError getRetrofitError() {
        return mRetrofitError;
    }

    /**
     * @return true if there are {@link ErrorDetails}
     * associated with this error. False otherwise.
     */
    public boolean hasErrorDetails() {
        return mErrorDetails != null;
    }

    /**
     * @return Details returned from the API associated with this error if present.
     */
    public ErrorDetails getErrorDetails() {
        return mErrorDetails;
    }
}

File: riot-api/src/main/java/fr/nspu/riot_api/RiotError.kt
package fr.nspu.riot_api

/**
 * Created by nspu on 09/03/18.
 */

import fr.nspu.riot_api.models.ErrorDetails
import fr.nspu.riot_api.models.ErrorResponse
import retrofit.RetrofitError

/**
 * This object wraps error responses from the API
 * and provides access to details returned by the request that are usually more
 * descriptive than default Retrofit error messages.
 *
 *
 * To use with asynchronous requests pass [RiotCallback]
 * instead of [retrofit.Callback] when making the request:
 */
class RiotError : Exception {

    val retrofitError: RetrofitError
    val errorDetails: ErrorDetails?

    constructor(retrofitError: RetrofitError, errorDetails: ErrorDetails?, message: String):super(message, retrofitError){
        this.retrofitError = retrofitError
        this.errorDetails = errorDetails
    }

    constructor(retrofitError: RetrofitError) : super(retrofitError){
        this.retrofitError = retrofitError
        this.errorDetails = null
    }

    /**
     * @return true if there are [ErrorDetails]
     * associated with this error. False otherwise.
     */
    fun hasErrorDetails(): Boolean = errorDetails != null


    companion object {

        fun fromRetrofitError(error: RetrofitError): RiotError {
            var errorResponse: ErrorResponse? = null

            try {
                errorResponse = error.getBodyAs(ErrorResponse::class.java) as ErrorResponse
            } catch (e: Exception) {
                e.printStackTrace()
            }

            if (errorResponse!!.error != null) {
                val message = errorResponse.error.status.toString() + " " + errorResponse.error.message
                return RiotError(error, errorResponse.error, message)
            } else {
                return RiotError(error)
            }
        }
    }
}
